File: Maximum Product Subarray.cpp
/*Given an array Arr that contains N integers (may be positive, negative or zero). 
Find the product of the maximum product subarray.
Input:
N = 5
Arr[] = {6, -3, -10, 0, 2}
Output: 180
Explanation: Subarray with maximum product
is  6, -3, -10 which gives product as 180.
*/

long long maxProduct(int * arr, int n) {
  long long product = arr[0];
  long long minimum = arr[0];
  long long maximum = arr[0];

  for (int i = 1; i < n; i++) {
    if (arr[i] < 0) {
      swap(minimum, maximum);
    }
    maximum = max((long long) arr[i], maximum * arr[i]);
    minimum = min((long long) arr[i], minimum * arr[i]);

    if (maximum > product) {
      product = maximum;
    }
  }
  return product;
}

File: Maximum Product Subarray.js
/* Given an integer array nums, find a contiguous non-empty subarray within the array that has the largest product, and return the product.

It is guaranteed that the answer will fit in a 32-bit integer.

A subarray is a contiguous subsequence of the array.

Input: nums = [2,3,-2,4]
Output: 6
Explanation: [2,3] has the largest product 6.
*/

var maxProduct = function(nums) {

    let max = Number.NEGATIVE_INFINITY
    for (let i = 0; i < nums.length; i++) {
        let product = 1
        for (let j = i; j < nums.length; j++) {
            product = product * nums[j]
            max = Math.max(product, max)
        }
    }
    return max
};
